presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java

```diff
@@ -138,7 +141,116 @@ public void testPushdownFilter()
                 plan -> assertTableLayout(plan, "lineitem", withColumnDomains(ImmutableMap.of(new Subfield("partkey", ImmutableList.of()), Domain.singleValue(BIGINT, 10L))), remainingPredicate, ImmutableSet.of("partkey", "orderkey")));
     }
 
-    private RowExpression constant(int value)
+    @Test
+    public void testSubfieldExtraction()
+    {
+        Session pushdownFilterEnabled = Session.builder(getQueryRunner().getDefaultSession())
+                .setCatalogSessionProperty(HIVE_CATALOG, PUSHDOWN_FILTER_ENABLED, "true")
+                .build();
+        assertUpdate("DROP TABLE IF EXISTS test_subfield_extraction");
+        assertUpdate("CREATE TABLE test_subfield_extraction(" +
+                         "id bigint, " +
+                         "a array(bigint), " +
+                         "b map(varchar, bigint), " +
+                         "c row(" +
+                             "a bigint, " +
+                             "b row(x bigint), " +
+                             "c array(bigint), " +
+                             "d map(bigint, bigint), " +
+                             "e map(varchar, bigint)))");
+        PlanMatchPattern test_subfield_extaction_scan = anyTree(tableScan("test_subfield_extraction"));
```

Comment:
- use lower-case camel-casing for variable names: testSubfieldExtractionScan
- anyTree(tableScan(...)) is too permissive; it doesn't ensure that there is no filter node on top of table scan; use `output(exchange(tableScan(tableName)))` instead
- consider extracting helper function to shorten the code below:

```
        assertPushdownFilter("SELECT * FROM test_subfield_extraction WHERE a[1] = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("a[1]"), Domain.singleValue(BIGINT, 1L)));

        assertPushdownFilter("SELECT * FROM test_subfield_extraction WHERE a[1 + 1] = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("a[2]"), Domain.singleValue(BIGINT, 1L)));

        assertPushdownFilter("SELECT * FROM test_subfield_extraction WHERE b['foo'] = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("b[\"foo\"]"), Domain.singleValue(BIGINT, 1L)));

        assertPushdownFilter("SELECT * FROM test_subfield_extraction WHERE b[concat('f','o', 'o')] = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("b[\"foo\"]"), Domain.singleValue(BIGINT, 1L)));

        assertPushdownFilter("SELECT * FROM test_subfield_extraction WHERE c.a = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("c.a"), Domain.singleValue(BIGINT, 1L)));

        assertPushdownFilter("SELECT * FROM test_subfield_extraction WHERE c.b.x = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("c.b.x"), Domain.singleValue(BIGINT, 1L)));

        assertPushdownFilter("SELECT * FROM test_subfield_extraction where c.c[5] = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("c.c[5]"), Domain.singleValue(BIGINT, 1L)));

        assertPushdownFilter("SELECT * FROM test_subfield_extraction where c.d[5] = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("c.d[5]"), Domain.singleValue(BIGINT, 1L)));

        assertPushdownFilter("SELECT * FROM test_subfield_extraction where c.e[concat('f', 'o', 'o')] = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("c.e[\"foo\"]"), Domain.singleValue(BIGINT, 1L)));

        assertPushdownFilter("SELECT * FROM test_subfield_extraction where c.e['foo'] = 1",
                "test_subfield_extraction",
                ImmutableMap.of(new Subfield("c.e[\"foo\"]"), Domain.singleValue(BIGINT, 1L)));
        ...
 
   private void assertPushdownFilter(String query, String tableName, Map<Subfield, Domain> predicateDomains)
    {
        assertPlan(pushdownFilterEnabled(), query,
                output(exchange(tableScan(tableName))),
                plan -> assertTableLayout(
                        plan,
                        tableName,
                        withColumnDomains(predicateDomains),
                        predicateDomains.keySet().stream().map(Subfield::getRootName).collect(toImmutableSet())));
    }

    private Session pushdownFilterEnabled()
    {
        return Session.builder(getQueryRunner().getDefaultSession())
                .setCatalogSessionProperty(HIVE_CATALOG, PUSHDOWN_FILTER_ENABLED, "true")
                .build();
    }
```

Reply:
Would it also make sense to remove repetition of the table name?